pkg/apis/triggers/v1alpha1/event_listener_defaults_test.go

```diff
@@ -0,0 +1,118 @@
+package v1alpha1_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/tektoncd/triggers/pkg/apis/triggers/v1alpha1"
+)
+
+func TestEventListenerSetDefaults(t *testing.T) {
+	tests := []struct {
+		name string
+		in   *v1alpha1.EventListener
+		want *v1alpha1.EventListener
+		wc   func(context.Context) context.Context
+	}{{
+		name: "with upgrade context - binding present",
+		in: &v1alpha1.EventListener{
+			Spec: v1alpha1.EventListenerSpec{
+				Triggers: []v1alpha1.EventListenerTrigger{{
+					DeprecatedBinding: &v1alpha1.EventListenerBinding{
+						Name: "my-binding",
+					},
+				}},
+			},
+		},
+		want: &v1alpha1.EventListener{
+			Spec: v1alpha1.EventListenerSpec{
+				Triggers: []v1alpha1.EventListenerTrigger{{
+					Bindings: []*v1alpha1.EventListenerBinding{{
+						Name: "my-binding",
+					}},
+				}},
+			},
+		},
+		wc: v1alpha1.WithUpgradeViaDefaulting,
+	}, {
+		name: "with upgrade context - no binding present",
+		in: &v1alpha1.EventListener{
+			Spec: v1alpha1.EventListenerSpec{
+				Triggers: []v1alpha1.EventListenerTrigger{{
+					Bindings: []*v1alpha1.EventListenerBinding{{
+						Name: "my-binding",
+					}},
+				}},
+			},
+		},
+		want: &v1alpha1.EventListener{
+			Spec: v1alpha1.EventListenerSpec{
+				Triggers: []v1alpha1.EventListenerTrigger{{
+					Bindings: []*v1alpha1.EventListenerBinding{{
+						Name: "my-binding",
+					}},
+				}},
+			},
+		},
+		wc: v1alpha1.WithUpgradeViaDefaulting,
+	}, {
+		name: "with upgrade context - both binding and bindings present",
+		in: &v1alpha1.EventListener{
+			Spec: v1alpha1.EventListenerSpec{
+				Triggers: []v1alpha1.EventListenerTrigger{{
+					DeprecatedBinding: &v1alpha1.EventListenerBinding{
+						Name: "my-binding",
+					},
+					Bindings: []*v1alpha1.EventListenerBinding{{
+						Name: "my-binding",
+					}},
```

Comment:
what if you have both but different names?

Reply:
Shouldn't matter. We just check if both fields exist; not what their names are

Reply:
Doesn't matter; we check if the field is present not what the name is!

Reply:
I thought that deprecated bindings were auto-magically internally re-written to `Bindings`, therefore if you specified something like:

```
						DeprecatedBinding: &v1alpha1.EventListenerBinding{
						Name: "my-same-binding",
					},
					Bindings: []*v1alpha1.EventListenerBinding{{
						Name: "my-same-binding",
					}},
```

Then the values in `Bindings` would be overwritten and you'd end up with only one binding? Which I'm thinking maybe should be an error? (I might be misunderstanding how this works tho!)

Reply:
Or, looking at the `want` field below, is the behavior that if you have both, the Deprecated field _isnt_ updated?

What about making this an error instead - i.e. if you use both, it's an error?

Reply:
Yeah, the flow is a bit confusing -- in the webhook setDefaults is called first, followed by the Validate. So, if both Binding and Bindings are present, SetDefaults does nothing, but the Validate function (which is called after SetDefaults) returns an Error. The end result is that the user gets a Validation error.

In the case where the user only has binding set (i.e. upgrading from 0.1 to 0.2), we'll move the binding value to bindings,

Reply:
Also, I think we'd have both binding and bindings set is if a user manually creates (or updates) a TriggerBinding to have both `binding` and `bindings`. In such cases, rather than silently moving the binding to bindings, I think returning a validation error is the way to go!